website/docs/opengl/c1/3-hello-texture.md

```diff
@@ -0,0 +1,462 @@
+        "Name": "1.3 - Hello Texture",
+        "Url": "3-hello-Texture.html",
+        "Metadata": {
+            "AuthorGitHub": "lumi2021",
+            "DateTimeWritten": "11/01/2024 16:00",
+            "PreviewImage": ""
+        }
+    }
+}
+---
+
+# 1.3 - Hello Texture
+<?# Info "You can view the source code for this tutorial [here](../sources/1.3-final-result.html). This tutorial builds on the previous tutorial. If you haven't read it, you can do so [here](1-hello-window.html)." /?>
+
+In the previous tutorial, we've shown you how to open an OpenGL window and draw a colored quad.
+                                                    
+Another very important thing in graphics programming is the ability to draw images. In this tutorial, you'll learn:
+
+* What a texture is.
+* How to send more data to shaders.
+* How to use texture coordinates.
+* How to import images as textures.
+* How shaders handle images.
+* How OpenGL handle transparency.
+* How to use texture parameters.
+* How to allow mipmaps.
+
+## What a texture is
+Technically, textures are multidimensional objects that store color data.
+
+The most common type of texture is the 2D texture, which stores a 2D grid of pixels of an image (this is just for comparison.
+Texture objects are a lot more complex than just 2D arrays!).
+This data is stored in GPU memory, and can be read a shader.
+
+<?# Info "There are multiple different dimensions of texture from 1D through to 3D. For the purposes of this tutorial, we will be focusing on 2D." /?>
+
+Textures are quite finicky to setup, and need quite a bit of information in order to be drawn.
+The most important piece of information that we need to send to the texture to see it on the screen are the texture coordinates.
+The texture coordinates are the positions within the texture for a particular vertex. Below we will learn how to fill out this data and read it from a shader.
+
+## Sending more data to shaders
+Texture coordinates are a type of data that can be manipulated by the vertex shader before it is used by the fragment shader.
+This means that this data needs to be passed to the vertex shader which, as per the previous tutorials, is once again done using a vertex buffer.
+
+Let's go back to our `OnLoad` function and update our `vertices` array to the following:
+```c#
+// The quad vertices data. Now with Texture coordinates!
+float[] vertices =
+{
+//       aPosition     | aTexCoords
+     0.5f,  0.5f, 0.0f,  1.0f, 1.0f,
+     0.5f, -0.5f, 0.0f,  1.0f, 0.0f,
+    -0.5f, -0.5f, 0.0f,  0.0f, 0.0f,
+    -0.5f,  0.5f, 0.0f,  0.0f, 1.0f
+};
+```
+
+With how we have structured the data, both the vertex and texture coordinates can be stored in a single array.
+
+If we try to run our program now...
+
+![A really messed up quad!](../../../images/opengl/chapter1/lesson3/messed-up-quad.png)
+
+Well, that doesn't like the quad we were expected. This is because we have updated the vertex buffer we're passing to the vertex shader,
```

Comment:
```suggestion
Well, that doesn't look like the quad we were expecting. This is because we have updated the vertex buffer we're passing to the vertex shader,
```

Reply:
Oops, my bad.